artio-core/src/main/java/uk/co/real_logic/artio/session/SessionParser.java

```diff
@@ -195,7 +196,11 @@ private Action rejectExceptionalMessage(final long messageType, final int refTag
     {
         if (messageType == LOGON_MESSAGE_TYPE)
         {
-            return onExceptionalMessage(logon.header(), refTagId, position);
+            final Action action = onExceptionalMessage(logon.header(), refTagId, position);
+            if (action != ABORT)
+            {
+                session.logoutAndDisconnect();
+            }
```

Comment:
This block should return the action, rather than fall-through. Otherwise in the case of back-pressure the first time you try to send the reject it will fall-through to onExceptionalMessage() call at the end of the method.